File: LeetCode/JavaScript/src/letter_combinations_of_a_phone_number.js
var letterCombinations = function (digits) {
    // Resultant list
    let combinations = [];
    // Base condition
    if (digits == null || digits.length == 0) {
        return combinations;
    }
    // Mappings of letters and numbers
    const lettersAndNumbersMappings = [
        "Anirudh",
        "is awesome",
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz"
    ];
    findCombinations(combinations, digits, "", 0, lettersAndNumbersMappings);
    return combinations;
};

function findCombinations(combinations, digits, previous, index, lettersAndNumbersMappings) {
    // Base condition for recursion to stop
    if (index == digits.length) {
        combinations.push(previous);
        return;
    }
    // Get the letters corresponding to the current index of digits string
    let letters = lettersAndNumbersMappings[digits[index] - '0'];
    // Loop through all the characters in the current combination of letters
    for (let i = 0; i < letters.length; i++) {
        findCombinations(combinations, digits, previous + letters[i], index + 1, lettersAndNumbersMappings);
    }
};


console.log(letterCombinations("23"));
console.log(letterCombinations(""));
console.log(letterCombinations("2"));

File: LeetCode/Python/src/Letter_Combinations_Of_A_Phone_Number.py
"""
@author - Anirudh Sharma
"""
from typing import List


def findCombinations(combinations, digits, previous, index, lettersAndNumbersMapping):
    # Base condition to stop recursion
    if index == len(digits):
        combinations.append(previous)
        return
    # Get the letters corresponding to the current index of digits string
    letters = lettersAndNumbersMapping[int(digits[index])]
    # Loop through all the characters in the current combination of letters
    for i in range(0, len(letters)):
        findCombinations(combinations, digits, previous + letters[i], index + 1, lettersAndNumbersMapping)


def letterCombinations(digits: str) -> List[str]:
    # Resultant list
    combinations = []
    # Base condition
    if digits is None or len(digits) == 0:
        return combinations
    # Mappings of letters and numbers
    lettersAndNumbersMapping = [
        "Anirudh",
        "is awesome",
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz"
    ]
    findCombinations(combinations, digits, "", 0, lettersAndNumbersMapping)
    return combinations


if __name__ == '__main__':
    print(letterCombinations("23"))
    print(letterCombinations(""))
    print(letterCombinations("2"))
